src/MagicWormhole/Internal/FileTransfer.hs

```diff
@@ -20,19 +20,28 @@ import Data.Aeson
   , object
   , withObject
   )
+import System.Posix.Types (FileOffset)
 
 -- | An offer made by a sender as part of the Magic Wormhole file transfer protocol.
 --
 -- Currently only supports sending simple text messages. A full version would
 -- also support sending files and directories.
-newtype Offer
+data Offer
   -- | A simple text message.
-  = Message Text deriving (Eq, Show)
+  = Message Text
+  | File FilePath FileOffset
+  deriving (Eq, Show)
 
 instance ToJSON Offer where
   toJSON (Message text) = object [ "offer" .= object [ "message" .= text ] ]
+  toJSON (File name size) = object [ "offer" .= object [ "file" .= object [ "filename" .= name, "filesize" .= fromEnum size ] ] ]
 
 instance FromJSON Offer where
   parseJSON = withObject "Offer" $ \obj -> do
     offer <- obj .: "offer"
-    Message <$> offer .: "message"
+    asum [ Message <$> offer .: "message"
+         , File
+           <$> ((offer .: "file") >>= (.: "filename"))
+           <*> (toEnum <$> ((offer .: "file") >>= (.: "filesize")))
+         ]
+
```

Comment:
I had some trouble resolving `asum`.  It doesn't seem to be imported.  And it's not in Prelude, right?  Then I wondered if it was in Protolude.  But the best I could manage was to find http://hackage.haskell.org/package/protolude-0.2.2/docs/doc-index-All.html which seems to say yes, it is in Protolude, but I was unable to find any more information about it.  So the best I could do was find `Data.Foldable.asum` - and maybe Protolude is just re-exporting that?

What am I missing?

From context I guess I see what's happening here, maybe.  either `Message ...` or `File ...` could succeed (possibly both if you had a malformed peer I suppose).  `Data.Foldable.asum` would then give back the first one that succeeds.  That sounds like the right thing to do here.

Reply:
`Protolude` exposes `asum` from `Data.Foldable`. https://github.com/sdiehl/protolude/blob/master/src/Protolude.hs (it qualifies the module as X and then reexports `X` at the top).